pkg/server/image_pull.go

```diff
@@ -26,195 +26,223 @@ import (
-		RepoTags:    []string{image},
-		RepoDigests: []string{digest},
-		Size:        uint64(size), // TODO(mikebrow):  compressed or uncompressed size? using compressed
-	}
-	if err = c.imageMetadataStore.Create(*meta); err != nil {
-		return &runtime.PullImageResponse{ImageRef: digest},
-			fmt.Errorf("pulled image `%q` but failed to store metadata for digest: %s err: %v", image, digest, err)
+		glog.V(4).Infof("PullImage using normalized image ref: %q", image)
 	}
 
-	// Return the image digest
-	return &runtime.PullImageResponse{ImageRef: digest}, err
-}
-
-// normalizeImageRef normalizes the image reference following the docker convention. This is added
-// mainly for backward compatibility.
-func normalizeImageRef(ref string) (string, error) {
-	named, err := reference.ParseNormalizedNamed(ref)
+	// TODO(random-liu): [P1] Schema 1 image is not supported in containerd now, we need to support
+	// it for backward compatiblity.
+	cfgDigest, digest, err := c.pullImage(ctx, image)
 	if err != nil {
-		return "", err
+		return nil, fmt.Errorf("failed to pull image %q: %v", image, err)
 	}
-	named = reference.TagNameOnly(named)
-	return named.String(), nil
-}
-
-// imageReferenceResolver attempts to resolve the image reference into a name
-// and manifest via the containerd library call..
-//
-// The argument `ref` should be a scheme-less URI representing the remote.
-// Structurally, it has a host and path. The "host" can be used to directly
-// reference a specific host or be matched against a specific handler.
-//
-// The returned name should be used to identify the referenced entity.
-// Dependending on the remote namespace, this may be immutable or mutable.
-// While the name may differ from ref, it should itself be a valid ref.
-//
-// If the resolution fails, an error will be returned.
-// TODO(mikebrow) add config.json based image.Config as an additional return value from this resolver fn()
-func (c *criContainerdService) imageReferenceResolver(ctx context.Context, ref string) (resolvedImageName string, manifest imagespec.Manifest, compressedSize uint64, err error) {
-	var (
-		size    int64
-		desc    imagespec.Descriptor
-		fetcher remotes.Fetcher
-	)
-
-	// Resolve the image name; place that in the image store; then dispatch
-	// a handler to fetch the object for the manifest
-	resolver := docker.NewResolver()
-	resolvedImageName, desc, fetcher, err = resolver.Resolve(ctx, ref)
-	if err != nil {
-		return resolvedImageName, manifest, compressedSize, fmt.Errorf("failed to resolve ref %q: err: %v", ref, err)
+	// Use config digest as imageID to conform to oci image spec.
+	// TODO(mikebrow): add truncIndex for image id
+	imageID := cfgDigest.String()
+	glog.V(4).Infof("Pulled image %q with image id %q, digest %q", image, imageID, digest)
+
+	var repoTag string
+	if _, ok := namedRef.(reference.NamedTagged); ok {
+		repoTag = namedRef.String()
+	}
+	repoDigest := getRepoDigest(namedRef, digest)
+	_, err = c.imageMetadataStore.Get(imageID)
```

Comment:
Offline discussed, we won't persist the metadata in the future. And we do use repodigests and repotags in cri-containerd when delete image now.